sympy/utilities/lambdify.py

```diff
@@ -386,21 +386,60 @@ def lambdify(args, expr, modules=None, printer=None, use_imps=True,
         for term in syms:
             namespace.update({str(term): term})
 
-    if _module_present('mpmath',namespaces) and printer is None:
+    if printer is None:
         #XXX: This has to be done here because of circular imports
-        from sympy.printing.lambdarepr import MpmathPrinter as printer
-
-    if _module_present('numpy',namespaces) and printer is None:
-        #XXX: This has to be done here because of circular imports
-        from sympy.printing.lambdarepr import NumPyPrinter as printer
-
-    if _module_present('numexpr',namespaces) and printer is None:
-        #XXX: This has to be done here because of circular imports
-        from sympy.printing.lambdarepr import NumExprPrinter as printer
+        wrap_printer = True
+        if _module_present('mpmath', namespaces):
+            from sympy.printing.pycode import MpmathPrinter as Printer
+        elif _module_present('numpy', namespaces):
+            from sympy.printing.pycode import NumPyPrinter as Printer
+        elif _module_present('numexpr', namespaces):
+            from sympy.printing.lambdarepr import NumExprPrinter as Printer
+            wrap_printer = False
+        elif _module_present('tensorflow', namespaces):
+            from sympy.printing.lambdarepr import TensorflowPrinter as Printer
+        elif _module_present('sympy', namespaces):
+            from sympy.printing.pycode import SymPyPrinter as Printer
+        else:
+            from sympy.printing.pycode import PythonCodePrinter as Printer
+
+        from sympy.matrices import MatrixBase
+
+        if wrap_printer:
+            class LambdifyPrinter(Printer):
+                """ Wrapper class for backward compatibility in lambdify """
+                def _print(self, expr):
+                    clsname = expr.__class__.__name__
+                    if clsname in namespace:
+                        func = namespace[clsname]
+                        if func.__module__ == 'mpmath.ctx_mp_python':
+                            self._module_format('mpmath.' + clsname)
+                        else:
+                            if func.__module__ and not '<' in func.__module__ and not '<' in func.__name__:
+                                try:
+                                    exec_('from %s import %s' % (func.__module__, func.__name__), {})
+                                except ImportError:
+                                    reg = False
+                                else:
+                                    reg = True
+                            else:
+                                reg = False
+                            self._module_format((
+                                (func.__module__ + '.') if func.__module__ else ''
```

Comment:
If you have a function within a function, like

```py
def test():
    def _test():
        pass

```

Then `_test.__module__` will be the module of the file it is defined in, but it won't be accessible from it.